tests/api/fuzzer.py

```diff
@@ -0,0 +1,319 @@
+                    print(f"FAIL: No default values in swagger for required parameter {name} in {self.url}")
+            for code,schema in spec['responses'].items():
+                if code=="200" and 'schema' in schema:
+                    self.success = schema['schema']
+                elif code=="200":
+                    L.error(f"200 reponse with no schema in {self.url}")
+                else:
+                    self.failures[code] = schema
+
+    def generate_success(self, host):
+        args = "&".join([f"{p.name}={some(p.values)}" for p in self.req_params.values() ])
+        base_url = urllib.parse.urljoin(host,self.url)
+        test_url = f"{base_url}?{args}"
+        if url_filter.match(test_url):
+            print(f"TEST: {test_url}")
+            return requests.get(url=test_url)
+        L.debug(f"url_filter skips {test_url}")
+        #for p in self.req_params.values():
+        #    p.dump()
+        #for p in self.opt_params.values():
+        #    p.dump()
+
+    def generate_failure(self, host):
+        args = "&".join([f"{p.name}={some(p.values)}" for p in self.req_params.values() ])
+        base_url = urllib.parse.urljoin(host,self.url)
+        test_url = f"{base_url}?{args}"
+        print(f"TEST: {test_url}")
+        #for p in self.req_params.values():
+        #    p.dump()
+        #for p in self.opt_params.values():
+        #    p.dump()
+        return requests.get(url=test_url)
+
+    def validate(self, resp, expect_success):
+        try:
+            resp_json = json.loads(resp.text)
+        except json.decoder.JSONDecodeError as e:
+            L.error("Non-json response - how to validate?")
+            return
+        if resp.status_code==200:
+            if self.success is not None:
+                does_response_fit_schema(posixpath.join(self.url,"200"), self.success, resp_json)
+            else:
+                L.error("Missing schema?")
+            #print(json.loads(resp.text))
+
+
+
+def get_the_spec(url):
+    if url[:7] == "file://":
+        with open(url[7:]) as f:
+            return f.read()
+    return requests.get(url=url).text
+
+# Swagger paths look absolute but they are relative to the base.
+def not_absolute(path):
+    return path[1:] if path[0] == '/' else path
+
+def find_ref(spec, path) :
+    if len(path)>0 and path[0] == '#':
+        return find_ref(spec,path[1:])
+    if len(path)==1:
+        return spec[path[0]]
+    return find_ref(spec[path[0]], path[1:])
+
+def resolve_refs(spec, spec_root=None):
+    '''Find all "$ref" keys in the swagger spec and inline their target schemas.
+
+       As with all inliners this will break if a definition recursively links to itself, but this should not
+       happen in swagger as embedding a structure inside itself would produce a record of infinite size.'''
+    if spec_root is None:
+        spec_root = spec
+    newspec = {}
+    for k,v in spec.items():
+        if k=="$ref":
```

Comment:
please, add space between variable, value and comparison here too.

Reply:
Do we have a style guide for python? I'm a fan of horizontal alignment to indicate structure in the expression so if we have not set one yet I have some ideas.